Sources/SmithyXML/Reader/Reader.swift

```diff
@@ -0,0 +1,317 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import typealias SmithyReadWrite.ReadingClosure
+import struct Foundation.Date
+import struct Foundation.Data
+import enum SmithyTimestamps.TimestampFormat
+import struct SmithyTimestamps.TimestampFormatter
+
+public class Reader {
+    public internal(set) var content: String?
+    public internal(set) var children: [Reader] = []
+    public internal(set) weak var parent: Reader?
+    public let nodeInfo: NodeInfo
+    public var nodeInfoPath: [NodeInfo] { (parent?.nodeInfoPath ?? []) + [nodeInfo] }
+
+    // MARK: - init & deinit
+
+    /// Creates an "empty" reader.  This Reader will be returned when the data cannot be parsed.
+    init() {
+        self.nodeInfo = ""
+    }
+
+    /// Used to create a new XML node during reading from XML.
+    /// - Parameter nodeInfo: The node info for this XML node.
+    init(nodeInfo: NodeInfo) {
+        self.nodeInfo = nodeInfo
+    }
+
+    private init(nodeInfo: NodeInfo, parent: Reader?) {
+        self.nodeInfo = nodeInfo
+        self.parent = parent
+    }
+
+    // MARK: - creating and detaching readers for subelements
+
+    public subscript(_ nodeInfo: NodeInfo) -> Reader {
+        if let match = children.first(where: {
+            nodeInfo.prefix == $0.nodeInfo.prefix && nodeInfo.name == $0.nodeInfo.name
+        }) {
+            return match
+        } else {
+            // The queried node doesn't exist.  Return one that has nil content.
+            return Reader(nodeInfo: nodeInfo, parent: self)
+        }
+    }
+
+    func addChild(_ child: Reader) {
+        children.append(child)
+        child.parent = self
+    }
+
+    /// Detaches this reader from its parent.  Typically used when this reader no longer
+    /// belongs in the tree, either because its data is nil or its contents were flattened
+    /// into its parents.
+    public func detach() {
+        parent?.children.removeAll { $0 === self }
+        parent = nil
+    }
+
+    public func unwrap() -> Reader {
+        let parent = Reader()
+        parent.addChild(self)
+        return parent
+    }
+
+    // MARK: - Reading values
```

Comment:
`.read()` and `.readIfPresent()` methods are provided for all data types that can be members of Smithy models.

`.read()` returns a non-nil value and throws if one cannot be constructed.  `.readIfPresent()` returns an optional.